### PR TITLE
feat: support dencun hard fork

### LINKED ISSUE #914: feat: EIP1153 (transient storage)
### Describe the Feature Request

The forthcoming merge introduces a new storage called transient storage, persistent during the whole life of a transaction but not committed to the state afterwards.

See https://eips.ethereum.org/EIPS/eip-1153

### Describe Preferred Solution

Much like SSTORE/SLOAD:

- add a `transient_storage` key in Account
- use this for TSTORE and TLOAD like SSTORE and SLOAD  use `storage`
- do no commit in `commit_accounts` (ie no need to update this).

> If a frame reverts, all writes to transient storage that took place between entry to the frame and the return are reverted, including those that took place in inner calls. This mimics the behavior of persistent storage.

### LINKED ISSUE #1000: feat: conform to EIP6780
### Describe the Feature Request

Previously we had `bug: `erase_contracts` doesn't completly wipe a contract` -> https://github.com/kkrt-labs/kakarot/issues/764
But with Dencun arriving, we don't need this feature anymore, we can bump to the business logic of

https://eips.ethereum.org/EIPS/eip-6780

### Describe Preferred Solution

We already have most of the features of this EIP as @enitrat pointed out in weekly.
In this timebox 1day, you will make sure that:
- We respect all the sub cases of the EIP (burn vs. transfer vs. delete etc. slots)

### LINKED ISSUE #1006: feat: support BLOBBASEFEE and other shard transactions opcodes for Dencun
### Describe the Feature Request

Dencung is coming soon (~2 weeks), it introduces a few new opcodes that we can't support functionally (unlike tstore, mcopy, etc. that we will support).

```[tasklist]
### Tasks
- [ ] add all the opcodes from dencun 
- [ ] mark them as unimplemented for now (not invalid opcodes)
```

### LINKED ISSUE #1028: epic: support cancun hard fork
### Describe the Feature Request

Support all the EVM-related EIPs included in the Cancun hard fork

- [ ] https://github.com/kkrt-labs/kakarot/issues/914
- [x] https://github.com/kkrt-labs/kakarot/issues/1000
- [x] https://github.com/kkrt-labs/kakarot/issues/1006
```[tasklist]
### Tasks
- [ ] https://github.com/kkrt-labs/kakarot/issues/914
- [x] https://github.com/kkrt-labs/kakarot/issues/1000
- [x] https://github.com/kkrt-labs/kakarot/issues/1006
- [ ] https://github.com/kkrt-labs/kakarot/issues/1038
- [ ] https://github.com/kkrt-labs/kakarot/issues/1039
```

### LINKED ISSUE #1038: dev: adapt debug script to support Pyspec tests
- Add a special behavior to handle test generation of Pyspec-based tests (different path architecture and naming)
- Skip validation of BEACON_ROOT storage

### LINKED ISSUE #1039: dev: support multi-transactions tests in debug scripts
Some tests verify the behavior of the system when some operations are split in different transactions (see: selfdestruct multi_tx tests)

As such, we need to adapt the debug script to support multi transactions

### LINKED ISSUE #1040: feat: add 0x10 as a valid precompile address, but revert txs to it
### If the feature request is approved, would you be willing to submit a PR? Yes

/ No _(Help can be provided if you need assistance submitting a PR)_

### PR DESCRIPTION
**Pull request by enitrat**
Time spent on this PR: 7d

## Pull request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Resolves #1000 
Resolves #1006
Resolves #1006
Resolves #1038
Resolves #1039 
Resolves #1040
Resolves #1028
Resolves #914 

## What is the new behavior?

-  Added the content of the Cancun hard fork.
- 
-

- - -
This change is [Reviewable](https://reviewable.io/reviews/kkrt-labs/kakarot/1027)

### CODE REVIEW

File: tests/end_to_end/PlainOpcodes/conftest.py [hunk 1]
```diff
@@ -1,7 +1,7 @@
 import pytest_asyncio
 
 
-@pytest_asyncio.fixture(scope="package")
+@pytest_asyncio.fixture
```
Comment (ClementWalter) [inline tests/end_to_end/PlainOpcodes/conftest.py:4]: this is going to redeploy a new counter each time, why did you remove this? (same for all contracts)

File: tests/end_to_end/PlainOpcodes/test_safe.py [hunk 2]
```diff
@@ -4,7 +4,7 @@
 from tests.utils.constants import ACCOUNT_BALANCE
 
 
-@pytest_asyncio.fixture(scope="package")
+@pytest_asyncio.fixture
```
Comment (ClementWalter) [inline tests/end_to_end/PlainOpcodes/test_safe.py:7]: same here, new contract every time?

File: tests/end_to_end/UniswapV2/conftest.py
Comment (ClementWalter) [inline tests/end_to_end/UniswapV2/conftest.py:1]: same here, new contract every time?

**Review by ClementWalter — COMMENTED** [review]
quite surprise by the feature scope update

File: tests/end_to_end/PlainOpcodes/conftest.py (hunk 1 above)
Comment (enitrat) [reply tests/end_to_end/PlainOpcodes/conftest.py:4]: otherwise this wouldn't run the tests at all. Would you have time to TAL ?

File: pyproject.toml [hunk 3]
```diff
@@ -40,7 +40,7 @@ pyperclip = "^1.8.2"
 tabulate = "^0.9.0"
 pytest-env = "^1.1.3"
 amarna = "^0.1.5"
-py-evm = "^0.7.0a1"
+py-evm = { git = "https://github.com/ethereum/py-evm", branch = "cancun-network-upgrade" }
```
[the remark this answers is not in the document]
Comment (enitrat) [reply pyproject.toml:43]: in `debug.py`, we use it to deserialize the block header RLP

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -168,6 +168,7 @@ jobs:
         uses: actions/checkout@v3
         with:
           repository: kkrt-labs/ef-tests
+          ref: feat/cancun
```
Comment (ClementWalter) [inline .github/workflows/ci.yml:171]: do we keep this?

File: kakarot_scripts/ef_tests/debug.py [hunk 5]
```diff
@@ -28,13 +28,16 @@
 RPC_ENDPOINT = "http://127.0.0.1:8545"
```
Comment (ClementWalter) [inline kakarot_scripts/ef_tests/debug.py:30]: too many blank lines

File: kakarot_scripts/ef_tests/fetch.py [hunk 6]
```diff
@@ -32,20 +32,37 @@ def generate_tests():
...
+    # Add tests from BlockchainTests/GeneralStateTests/Pyspecs
+    test_cases.update(
+        {
+            f"{name.split('::')[-1]}": content
+            for (root, _, files) in os.walk(EF_TESTS_DIR)
+            for file in files
+            if file.endswith(".json")
+            and "BlockchainTests/GeneralStateTests/Pyspecs" in root
+            for name, content in json.loads((Path(root) / file).read_text()).items()
+        }
```
Comment (ClementWalter) [inline kakarot_scripts/ef_tests/fetch.py:51]: no need to filter on fork here?

File: kakarot_scripts/ef_tests/fetch.py [hunk 7]
```diff
@@ -32,20 +32,37 @@ def generate_tests():
...
+        and "BlockchainTests/GeneralStateTests" in root
+        and "Pyspecs" not in root
         for name, content in json.loads((Path(root) / file).read_text()).items()
         if content.get("network") == DEFAULT_NETWORK
     }
 
+    # Add tests from BlockchainTests/GeneralStateTests/Pyspecs
+    test_cases.update(
+        {
+            f"{name.split('::')[-1]}": content
```
Comment (ClementWalter) [inline kakarot_scripts/ef_tests/fetch.py:45]: the name convention changes then?

File: kakarot_scripts/ef_tests/fetch.py [hunk 8]
```diff
@@ -32,20 +32,37 @@ def generate_tests():
...
     shutil.rmtree(EF_TESTS_PARSED_DIR, ignore_errors=True)
     EF_TESTS_PARSED_DIR.mkdir(parents=True, exist_ok=True)
 
     for test_name, test_case in test_cases.items():
-        json.dump(
-            test_case,
-            open(EF_TESTS_PARSED_DIR / f"{test_name}.json", "w"),
-            indent=4,
-        )
+        try:
```
Comment (ClementWalter) [inline kakarot_scripts/ef_tests/fetch.py:58]: what would make this fail?

File: src/backend/starknet.cairo [hunk 9]
```diff
@@ -162,9 +166,9 @@ namespace Internals {
     // @param self The pointer to the Account
     // @param starknet_address A starknet address to commit to
     // @notice Iterate through the storage dict and update the Starknet storage
-    func _commit_account{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
-        self: model.Account*
-    ) {
+    func _commit_account{
```
Comment (ClementWalter) [inline src/backend/starknet.cairo:169]: update docstring

File: tests/src/kakarot/precompiles/test_precompiles.py [hunk 10]
```diff
@@ -10,6 +12,7 @@ class TestRun:
                 (0x6, "Kakarot: NotImplementedPrecompile 6"),
                 (0x7, "Kakarot: NotImplementedPrecompile 7"),
                 (0x8, "Kakarot: NotImplementedPrecompile 8"),
+                (0x0A, "Kakarot: NotImplementedPrecompile :"),
```
Comment (ClementWalter) [inline tests/src/kakarot/precompiles/test_precompiles.py:15]: plz update the Error.notImplementedPrecomile(address) to handle 10 (here 10 becomes `:` because in ASCII, after `9` it's `:`)

File: tests/src/kakarot/test_state.py [hunk 11]
```diff
@@ -89,8 +89,8 @@ class TestCachePreaccessedAddresses:
         @SyscallHandler.patch("IERC20.balanceOf", lambda addr, data: [0, 1])
         def test_should_cache_precompiles(self, cairo_run):
             state = cairo_run("test__cache_precompiles")
-            assert list(state["accounts"].keys()) == [
-                f"0x{i:040x}" for i in range(1, 10)
+            assert list(map(str.lower, state["accounts"].keys())) == [
+                f"0x{i:040x}" for i in range(1, 11)
```
Comment (ClementWalter) [inline tests/src/kakarot/test_state.py:93]: easier to just do `f"0x{i:040x}".upper()`
but the truth is to use `to_checksum_address`

**Review by ClementWalter — CHANGES_REQUESTED** [review]
few minor changes to polish the PR

File: kakarot_scripts/ef_tests/fetch.py (hunk 6 above)
Comment (enitrat) [reply kakarot_scripts/ef_tests/fetch.py:51]: indeed, we can filter with `if f"fork_{DEFAULT_NETWORK}" in name` because pyspec tests have the fork in the name, like `src/test_access_list[fork_Cancun-blockchain_test]`

File: kakarot_scripts/ef_tests/fetch.py (hunk 7 above)
Comment (enitrat) [reply kakarot_scripts/ef_tests/fetch.py:45]: Pyspec tests have different names, e.g. `src/GeneralStateTestsFiller/Pyspecs/berlin/eip2930_access_list/test_acl.py::test_access_list[fork_Berlin-blockchain_test]` instead of the regular `access_list_d0g0v0` for non-python based tests

For them

File: kakarot_scripts/ef_tests/fetch.py (hunk 8 above)
Comment (enitrat) [reply kakarot_scripts/ef_tests/fetch.py:58]: filename too longs :(
```
[Errno 63] File name too long: 'tests/ef_tests/test_data/parsed/test_selfdestruct_created_same_block_different_tx[fork_Cancun-blockchain_test-selfdestruct_contract_address_0x6295ee1b4f6dd65047762f924ecd367c17eabf8f-entry_code_address_0xec0e71ad0a90ffe1909d27dac207f7680abba42d-call_times_1-selfdestruct_contract_initial_balance_0].json'
[Errno 63] File name too long: 'tests/ef_tests/test_data/parsed/test_selfdestruct_created_same_block_different_tx[fork_Cancun-blockchain_test-selfdestruct_contract_address_0x6295ee1b4f6dd65047762f924ecd367c17eabf8f-entry_code_address_0xec0e71ad0a90ffe1909d27dac207f7680abba42d-call_times_1-selfdestruct_contract_initial_balance_1].json'
[Errno 63] File name too long: 'tests/ef_tests/test_data/parsed/test_selfdestruct_created_same_block_different_tx[fork_Cancun-blockchain_test-selfdestruct_contract_address_0x6295ee1b4f6dd65047762f924ecd367c17eabf8f-entry_code_address_0xec0e71ad0a90ffe1909d27dac207f7680abba42d-call_times_10-selfdestruct_contract_initial_balance_0].json'
[Errno 63] File name too long: 'tests/ef_tests/test_data/parsed/test_selfdestruct_created_same_block_different_tx[fork_Cancun-blockchain_test-selfdestruct_contract_address_0x6295ee1b4f6dd65047762f924ecd367c17eabf8f-entry_code_address_0xec0e71ad0a90ffe1909d27dac207f7680abba42d-call_times_10-selfdestruct_contract_initial_balance_1].json'
```

not sure how to fix this efficiently

Comment (enitrat) [reply kakarot_scripts/ef_tests/fetch.py:58]: given that the error is very restricted to these tests only and that we don't need to debug them I don't think it's a problem

File: tests/src/kakarot/precompiles/test_precompiles.py (hunk 10 above)
Comment (enitrat) [reply tests/src/kakarot/precompiles/test_precompiles.py:15]: added a special handling of case == 10 (only precompile w/ two digits)

File: tests/src/kakarot/test_state.py (hunk 11 above)
Comment (enitrat) [reply tests/src/kakarot/test_state.py:93]: done

File: tests/src/kakarot/precompiles/test_precompiles.py (hunk 10 above)
Comment (enitrat) [reply tests/src/kakarot/precompiles/test_precompiles.py:15]: added a special handling of case == 10 (only precompile w/ two digits)

File: .github/workflows/ci.yml (hunk 4 above)
Comment (enitrat) [reply .github/workflows/ci.yml:171]: it was merged now

File: kakarot_scripts/ef_tests/fetch.py (hunk 8 above)
Comment (ClementWalter) [reply kakarot_scripts/ef_tests/fetch.py:59]: looks like the naming is directly produced by a `pytest.mark.parametrize` would could replace the `[]` content with numbers?

File: src/kakarot/errors.cairo [hunk 12]
```diff
@@ -428,7 +428,13 @@ namespace Errors {
         assert [error + 31] = 108;  // l
         assert [error + 32] = 101;  // e
         assert [error + 33] = 32;  //
-        assert [error + 34] = '0' + address;  //
+
+        if (address == 10) {
```
Comment (ClementWalter) [inline src/kakarot/errors.cairo:432]: pragmatique 🫡

File: docs/general/decode_a_cairo_trace.md [hunk 13]
```diff
@@ -68,12 +68,12 @@ this cryptic message to map location in the original cairo program.
 And tada! You should now be able to understand the reason your cairo program
 failed with more clarity.
 
-## Extended scripts
+## Extended kakarot_scripts
```
[the remark this answers is not in the document]
Comment (ClementWalter) [reply docs/general/decode_a_cairo_trace.md:71]: @enitrat here

File: kakarot_scripts/ef_tests/fetch.py (hunk 8 above)
Comment (enitrat) [reply kakarot_scripts/ef_tests/fetch.py:59]: I could theoritically but then it becomes completely useless to debug from the test name directly as if the EF-Test fails, it will fail with this parametrized name

**Review by ClementWalter — APPROVED** [review]
